### ISSUE TITLE
Wildcard and select() not working on property matching

### ISSUE THREAD

**Issue by moifort** [gvergnaud/ts-pattern#40]
**Describe the bug**
Wildcard `__` and `select()` not working on property matching. As a workaround I use `__.number` 

**Code Sandbox with a minimal reproduction case**
Result on console

https://codesandbox.io/s/nervous-turing-8cffw?file=/src/index.ts:0-937

**Versions**
- TypeScript version: 4.2.3
- ts-pattern version: 3.2.1
- environment: node 14

**Sample code**

```ts
import { match, select, __ } from "ts-pattern";

type Id = { teamId: number } | { storeId: number };

const selectedId: Id = { teamId: 1 };

// Bug with wildcard
const bugWildcard = match<Id>(selectedId)
  .with({ storeId: __ }, () => "storeId")
  .with({ teamId: __ }, () => "teamId")
  .exhaustive(); // Return storeId instead of teamId

console.log(
  "WILDCARD BUG",
  "match result:",
  bugWildcard,
  "value to match:",
  selectedId
);

// Bug with select
const value = match<Id>(selectedId)
  .with({ storeId: select() }, (storeId) => storeId)
  .with({ teamId: select() }, (teamId) => teamId)
  .exhaustive(); // Return nothing

console.log(
  "SELECT BUG",
  "match result:",
  value,
  "value to match:",
  selectedId
);

// Working with __.number
const number = match<Id>(selectedId)
  .with({ storeId: __.number }, () => "storeId")
  .with({ teamId: __.number }, () => "teamId")
  .exhaustive();

console.log("WORKING", "match result:", number, "value to match:", selectedId);


```

**Comment by gvergnaud** [comment]
I think what's happening is that since `__` and `select()` always match, if you try to match on a property that doesn't exist on an object, it's actually `undefined` which matches the pattern. 

It would probably make more sense if `{ abc: __ }` was matching only if the `abc` key exists on the input, though. Is this what you were expecting?

I'll have a look at the TC39 pattern matching proposal to check how the plan to handle this case

**Comment by moifort** [comment]
@gvergnaud thanks for your quick answer!

**Comment by gvergnaud** [comment]
Fixed in v3.2.2!
